Python/ceval.c

```diff
@@ -4066,7 +4066,88 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
             CHECK_EVAL_BREAKER();
             DISPATCH();
         }
-
+        case TARGET(CALL_FUNCTION_ADAPTIVE): {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            if (cache->adaptive.counter == 0) {
+                next_instr--;
+                if (_Py_Specialize_CallFunction(BUILTINS(), stack_pointer,
+                    cache->adaptive.original_oparg, next_instr, cache) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(CALL_FUNCTION, deferred);
+                cache->adaptive.counter--;
+                oparg = cache->adaptive.original_oparg;
+                JUMP_TO_INSTRUCTION(CALL_FUNCTION);
+            }
+        }
+        case TARGET(CALL_CFUNCTION_FAST): {
+            /* Builtin METH_FASTCALL functions, without keywords */
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyCallCFunctionCache *cache1 = &caches[-1].call_function;
+            PyObject **callable = &PEEK(cache0->original_oparg + 1);
+            DEOPT_IF(!PyCFunction_CheckExact(*callable), CALL_FUNCTION);
+            PyCFunction cfunc = PyCFunction_GET_FUNCTION(*callable);
+            DEOPT_IF(cfunc != cache1->cfunc, CALL_FUNCTION);
+
+            assert(cframe.use_tracing == 0);
+            // res = func(self, args, nargs)
+            PyObject *res = ((_PyCFunctionFast)(void(*)(void))cfunc)(
+                PyCFunction_GET_SELF(*callable),
+                stack_pointer - cache0->original_oparg,
+                cache0->original_oparg);
+            assert((res != NULL) ^ (_PyErr_Occurred(tstate) != NULL));
+
+            /* Clear the stack of the function object. */
+            while (stack_pointer > callable) {
+                PyObject *x = EXT_POP(stack_pointer);
+                Py_DECREF(x);
+            }
+            PUSH(res);
+            if (res == NULL) {
+                /* Not deopting because this doesn't mean our optimization was wrong.
```

Comment:
In which case the `record_cache_hit()` and `STAT_INC()` should come before this if statement.